[WIP] Jenkins tests on Windows.

syclik opened the pull request:
## Summary
This PR enables Windows unit tests. (This is being submitted now to kick off the build on Jenkins.)

The goal is to fix all issues with Windows before this is merged.

## Tests


## Side Effects


## Checklist

- [x] Math issue #951

- [x] Copyright holder: Generable

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

rok-cesnovar commented:
[These ](https://github.com/stan-dev/math/pull/1046/commits/cf1a9e0a6893107008cd66459edebd72c141692a)are the changes I had to do in order to make things run and pass on Windows. It seems that `matrix_exp_computeUV::run` is having problems with stan::math::var Eigen matrices on Windows. The current solution is just a quick fix to see if this also works on CI. We need to find a more general solution probably.

@yizhang-cae can you please verify that this fix is actually OK.

syclik commented:
The first problem I'm seeing is:
- [ ] `test/unit/math/prim/mat/err/check_pos_definite_test.exe`

rok-cesnovar commented:
Oh, yeah. 

This test causes the Windows Error Reporting popup on the EXPECT_DEATH, because the child process crashes as part of this death test. You need to manually cancel the popup and then the test continues and passes the tests. I did a quick fix based on [this](https://github.com/google/googletest/pull/1120) just to see if this fixes it. Should we upgrade the gtest lib?

syclik commented:
Wow, you're fast! I was trying to dig into how to deal with the problem and
still searching whether it works on Windows.

Yes, I think we should update Google Test. It makes sense to be an
independent issue + PR. Want to get that started? Or I can write up the
issue if that helps.

On Tue, Oct 16, 2018 at 1:02 PM Rok Češnovar <notifications@github.com>
wrote:

> Oh, yeah.
>
> This test causes the Windows Error Reporting popup on the EXPECT_DEATH,
> because the child process crashes as part of this death test. You need to
> manually cancel the popup and then the test continues and passes the tests.
> I did a quick fix based on this
> <https://github.com/google/googletest/pull/1120> just to see if this
> fixes it. Should we upgrade the gtest lib?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1046#issuecomment-430316478>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_F-4tRCasVvMfviKXbRXGX7cf2Fb8ks5ulhEggaJpZM4XASPZ>
> .
>

rok-cesnovar commented:
Yeah, no problem. I will do the issue tonight. The PR will have to wait till tomorrow, I have to finish up the docs stuff for the latest GPU PR.

rok-cesnovar commented:
After merging the gtest changes, tests now fail at the matrix_exp_multiply test. 
I have written down the stuff I narrowed down so far on the seg fault in the discourse thread: https://discourse.mc-stan.org/t/the-math-library-is-not-passing-tests-on-windows-this-is-the-next-thing-im-working-on/5989/12?u=rok_cesnovar

syclik commented:
Thanks! I'll set aside time to jump in this week.

On Mon, Oct 29, 2018 at 3:46 PM Rok Češnovar <notifications@github.com>
wrote:

> After merging the gtest changes, tests now fail at the matrix_exp_multiply
> test.
> I have written down the stuff I narrowed down so far on the seg fault in
> the discourse thread:
> https://discourse.mc-stan.org/t/the-math-library-is-not-passing-tests-on-windows-this-is-the-next-thing-im-working-on/5989/12?u=rok_cesnovar
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1046#issuecomment-434052605>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_FxlBxNdR89wjTyNHfOKRnhkXPLbzks5up1sCgaJpZM4XASPZ>
> .
>

syclik commented:
Ok... there's a fix for the segfault in `matrix_exp_multiply`. Hopefully that's it and we can get things testing on Windows again.

seantalts commented:
Thanks all! Taking a look and trying to get my bearings. Have a few questions:
> ad_promotable had to be updated to include long double
Seems like it had long double before, here:
https://github.com/stan-dev/math/pull/1046/files#diff-0b6a929f9f467e93bc45dd154b930bfaL41
A lot of the promotable things seem to have been reordered, potentially? Just to double check, the order doesn't matter for semantics does it? Could someone possibly summarize what that bit is about for me? The rest of the changes look good.

syclik commented:
@seantalts... I was incorrect. We had to add: `long long` and `unsigned long long`. (I just combed through all the changes and those were the two new ones.) I'll update the top-level comment.

I'm not sure why it was reordered, but the order shouldn't matter for semantics. Since we're being specific, the ADL rules should guarantee that it is able to promote.

It looks like there are 0 tests for `ad_promotable`. If you think it'd help, we should add them.

syclik commented on stan/math/prim/scal/meta/ad_promotable.hpp:
```diff
@@ -107,26 +108,42 @@ struct ad_promotable<unsigned long, double> {  // NOLINT(runtime/int)
 
 /**
- * An unsigned int may be promoted to a double.
+ * A long long may be promoted to a double.
```
@seantalts: `long long` and `unsigned long long` are the two new specializations. (If you remove these two and try to run on Windows, it'll fail to compile.)

seantalts commented on test/unit/math/rev/mat/prob/bernoulli_logit_glm_lpmf_test.cpp:
```diff
@@ -37,7 +38,7 @@ TEST(ProbDistributionsBernoulliLogitGLM,
     Matrix<int, Dynamic, 1> y(3, 1);
     for (size_t i = 0; i < 3; i++) {
-      y[i] = Matrix<uint, 1, 1>::Random(1, 1)[0] % 2;
+      y[i] = abs(Matrix<int, 1, 1>::Random(1, 1)[0] % 2);
```
Does this not work with `unsigned int` (instead of `uint`)? If it does, maybe we should test that instead?

syclik replied:
I just checked... it does work with `unsigned int` and I changed all the instances with that so it looks like the old code.

seantalts commented:
Gotcha. Tests would be helpful but I think that this PR will leave the codebase better than we found it, so I'm happy to let it go through as-is - just one more question about the abs vs. unsigned int in the tests.

syclik commented:
@seantalts. I just added tests for `ad_promotable`. (It would have helped to have this test prior to the PR to isolate the issue, so I think it's good we add it now.)

syclik commented:
@seantalts: updates since your review:
1. Added tests for `ad_promotable` for prim, rev, fwd, and mix.
2. Fixed a small typo for `ad_promotable` in prim
3. Changed the `abs(Matrix<int>)` to `Matrix<unsigned int>`. I agree that it's cleaner that way (and it mirrors the old code better.

syclik commented:
@seantalts, mind reviewing again? The last changes were due to a mistake I made when updating from `unit` to `unsigned int` (then dealing with linting after that).

syclik commented:
This commit fixes the error I introduced: https://github.com/stan-dev/math/pull/1046/commits/2c618de4022a03ae818a3d7da2dff295b540be24
